service_container/service_locators.rst

```diff
@@ -0,0 +1,154 @@
+.. index::
+    single: DependencyInjection; Service Locators
+
+Service Locators
+================
+
+What is a Service Locator
+-------------------------
+
+Sometimes, a service needs the ability to access other services without being sure
+that all of them will actually be used.
+
+In such cases, you may want the instantiation of these services to be lazy, that is
+not possible using explicit dependency injection since services are not all meant to
+be ``lazy`` (see :doc:`/service_container/lazy_services`).
+
+Service Locators are intended to solve this problem by giving access to a set of
+identified services while instantiating them only when really needed, providing
+laziness.
+
+A real-world example being a CommandBus which maps CommandHandler services by Command
+class names and use them to handle their respective command when it is asked for::
+
+    // ...
+    class CommandBus
+    {
+        /**
+         * @var CommandHandler[]
+         */
+        private $handlerMap;
+
+        // ...
+
+        public function handle(Command $command)
+        {
+            foreach (handlerMap as $commandClass => $handler) {
```

Comment:
No need for iterating over the handler map, right?

```php
$commandClass === get_class($command);
if (isset($handlerMap[$commandClass])) {
    $handler = $handlerMap[$commandClass];

    return $handler->handle($command);
}
```

(or even using null coalesce operator) ? 😄